### PR TITLE
Feature: Add Block Editor Settings and UI tweaks

### PR DESCRIPTION
**Pull request by geriux**
**Related PRs:**

- https://github.com/wordpress-mobile/WordPress-iOS/pull/23499

## Description

This PR adds the following core functionality to the editor:

### Block Editor Settings

It adds the editor's settings data, currently by fetching the block editor settings endpoint [within the editor](https://github.com/wordpress-mobile/GutenbergKit/pull/21/files#diff-11987f198cb9cf62a83bd4322a3e37df80cecd4ce6f20a5a72038a7f93c0905dR26).

This is a temporary approach, as this data should be provided by the host apps.

This is needed to obtain important data like `allowedMimeTypes`, theme `styles`, and other settings required for the editor to function fully.

### Loading Post Content in the Editor

It adds some post type entities to allow a post to be loaded by passing the content without fetching it from the server. This is similar to the approach [we used with the React Native editor](https://github.com/WordPress/gutenberg/blob/20f7dfd8a68b0db7965f12d165f6f628b6c4b68f/packages/editor/src/components/provider/index.native.js#L47).

After adding the entities, it uses `receiveEntityRecords` with the post data from the host apps, if available.

###  Media and Patterns Support

With the block editor settings, media uploads are now enabled, as well as fetching the available patterns and reusable blocks.

### EditorToolbar

Updates this component to remove unused code and improve functionality by reusing Gutenberg components.

### Usage of Experimental APIs

You'll see some usage of experimental components by unlocking some private editor APIs. These are currently used in the web editor, so there shouldn't be any issues.

### Style Updates

This PR also updates several style issues related to the editor, the inserter, and the blocks.

### Native Side

Adds several new parameters to initialize the WebView. Most of these are related to the post data and editor settings.

I also commented on some native code that needs refactoring or complete removal.

## Testing

> [!NOTE]  
> To be able to test the editor please enable the feature flag `Experimental Block Editor` by going to Settings > App Settings > Debug > Feature Flags
>
> Please use the following build:
> - [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/23499#issuecomment-2289528270)

## Media Library

**Images**

- Open the editor
- Open the inserter
- Go to the Media tab
- **Expect** to see the `Images` list item
- Go to Images
- **Expect** to see images from your media library
- Tap on an image
- **Expect** the image to be added to the post

**Openverse**

- Open the editor
- Open the inserter
- Go to the Media tab
- **Expect** to see the `Openverse` list item
- Go to Openverse
- **Expect** to see images from the Openverse directory
- Tap on an image
- **Expect** the image to be uploaded and added to the post

## Patterns

**Patterns Library**

- Open the editor
- Open the inserter
- Go to the Patterns tab
- **Expect** to see the Patterns library list
- Go to any section
- Tap on any Pattern
- **Expect** the Pattern to be added to the post

**Custom patterns**

- Open the editor
- Open the inserter
- Go to the Patterns tab
- **Expect** to see the `My patterns` item
- Tap on `My Patterns`
- **Expect** to see your patterns
- Tap on one of the patterns
- **Expect** the pattern to be added to the post

## Opening a Post

- Go to the Posts list
- Tap on an existing post
- **Expect** to see the editor with the post tile and content

## Inserter

- Open the editor
- Tap on the (+) button to open the inserter
- Scroll through the list (before some events affected the scroll)
- Tap on any block
- **Expect** to add a block by just tapping once (before you had to tap twice)

**Note**: Some hover styles were removed as well.

## Theme Styles

- Open the app
- Go to Settings > App Settings > Debug > Feature Flags
- Scroll down to the bottom and enable `Experimental Block Editor Styles`
- Open the editor
- **Expect** to see your theme styles in the editor's canvas

**Note:** There will be a bit of delay while the endpoint with the style settings fetches the data. This should be fixed once this data comes from the host app.

### Media

https://github.com/user-attachments/assets/f99e0b10-8097-409c-a7f0-dd1708351805

### Patterns

https://github.com/user-attachments/assets/f14c4436-3445-4f15-8e12-0ff88ef97389

### Opening a Post

https://github.com/user-attachments/assets/fbecc984-b7bb-4ca0-9278-f847fb6aedf2

### Inserter

https://github.com/user-attachments/assets/50c30271-75c2-4b11-b21b-64f79372958e

### Theme Styles

https://github.com/user-attachments/assets/55d9e3c4-662d-4881-92f8-5540c8aafd80

### CODE REVIEW

File: ReactApp/src/components/Editor.jsx
Comment (geriux) [inline ReactApp/src/components/Editor.jsx:116]: I'll work on a different PR to provide post updates to the host apps

File: ReactApp/src/components/hooks/use-editor-styles.js
Comment (geriux) [inline ReactApp/src/components/hooks/use-editor-styles.js:15]: Or when this is part of Gutenberg and we can import it directly

File: ReactApp/src/components/Editor.jsx
Comment (dcalhoun) [inline ReactApp/src/components/Editor.jsx:66]: The `editor` assignment feels like unnecessary indirection to me. We might consider removing it and referencing `window.editor` elsewhere directly instead. 

I recognize `editor` was present before the proposed changes. Merely capturing my thought, we can address it later, if desired.

File: ReactApp/src/components/EditorToolbar.jsx
Comment (dcalhoun) [inline ReactApp/src/components/EditorToolbar.jsx]: For the pursuit of brevity, we might consider self closing this element, given it has no children. 
```suggestion
						/>
```

File: ReactApp/src/components/hooks/use-editor-styles.js
Comment (dcalhoun) [inline ReactApp/src/components/hooks/use-editor-styles.js]: To preserve hot module reloading, we might consider adhering to this lint warning and relocating the export to a separate file.

File: ReactApp/src/index.css
Comment (dcalhoun) [inline ReactApp/src/index.css:124]: There are a few[ "magic numbers"](https://en.wikipedia.org/wiki/Magic_number_(programming)) in the styles currently. In a separate PR later, I posit we should make a pointed effort to explain them through variable names, design systems, code comments, etc.

File: ReactApp/src/index.css
Comment (dcalhoun) [inline ReactApp/src/index.css:28]: At some point in the future, we should take time to consider how we might best coexist with core's styles. I.e., the use of `!important` concerns me, it is likely worthy of exploring a more sustainable override approach. 

Not a task for this PR, merely sharing the thought. Related to https://github.com/Automattic/dotcom-forge/issues/8678.

File: ReactApp/src/main.jsx
Comment (dcalhoun) [inline ReactApp/src/main.jsx]: Given the close proximity to the `window.GBKit` assignment, it likely has little impact, but I suggest we always rely upon the `getGBKit` utility. The consistency in approach might mitigate unexpected, cryptic errors elsewhere if `window.GBKit` is referenced directly. WDYT? 

```suggestion
	const { themeStyles, siteApiRoot } = getGBKit();
```

File: ReactApp/src/misc/Helpers.js
Comment (dcalhoun) [inline ReactApp/src/misc/Helpers.js]: I envisioned all globals residing under a single `GBKit` namespace to avoid potential collisions. 

```js
const { initialPost: currentPost } = getGBKit();
```

Do you feel that is valuable?

File: ReactApp/src/misc/Helpers.js
Comment (dcalhoun) [inline ReactApp/src/misc/Helpers.js]: Given the wrapping conditional, this optional chain is unnecessary, correct? Should we instead rely upon the OR operator? 
```suggestion
			type: currentPost.type || 'post',
```

File: ReactApp/src/misc/api-fetch-setup.js
Comment (dcalhoun) [inline ReactApp/src/misc/api-fetch-setup.js:95]: In what context is the namespace already present?

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [inline Sources/GutenbergKit/Sources/EditorViewController.swift:43]: I feel we are approaching an argument count threshold where we should consider an alternative pattern — builder pattern, parameter object, property setters, etc. We can address this separately, if desired.

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [inline Sources/GutenbergKit/Sources/EditorViewController.swift]: I believe this might be simplified. WDYT?  
```suggestion
        let postId = id ?? nil
        let hasThemeStylesEnabled = themeStyles ?? false
```

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [inline Sources/GutenbergKit/Sources/EditorViewController.swift]: Does `postId` serve a specific purpose or could we rely upon `id` directly?

**Review by dcalhoun — APPROVED** [review]
Great work! This is a lot of crucial functionality that is incredibly valuable to merge. 

Testing went relatively well for me. The two things I noted: 

1. Selecting media from Media Library does not work for me; tapping the button doesn't display anything. 
2. Media uploaded to private sites results in broken images within the editor. 

The first might be worth debugging now, the latter we could address later. 

I left several comments that I consider subjective and refactors that might be nice to have. However, I lean towards merging this sooner rather than later, as it overlaps with my own work and unlocks a lot of capabilities. We could follow up with improvements later.

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: Actually, we likely do not even need `let postId = id ?? nil`. We could just pass `id`.

File: ReactApp/src/index.css
Comment (geriux) [reply ReactApp/src/index.css:124]: I agree! To be honest this whole file needs to be updated and possibly split. I think we can work on that separately addressing other issues like the `!important` usage mentioned in this [comment](https://github.com/wordpress-mobile/GutenbergKit/pull/21#discussion_r1744350860).

File: ReactApp/src/index.css
Comment (geriux) [reply ReactApp/src/index.css:28]: Definitely! I think once we have a working prototype with the needed features we should spend some time on that. Take some time to clean up and improve what we have now.

File: ReactApp/src/misc/Helpers.js
Comment (geriux) [reply ReactApp/src/misc/Helpers.js]: Let's try it! It'd be a good idea to have all of the editor params in the same namespace/localStorage 👍

File: ReactApp/src/misc/api-fetch-setup.js
Comment (geriux) [reply ReactApp/src/misc/api-fetch-setup.js:95]: I'll test it again and report back, but I remember I started to have issues once we merged this code and I was playing around with some endpoints, where it was appending the namespace at least twice 😅

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (geriux) [reply Sources/GutenbergKit/Sources/EditorViewController.swift:43]: Yeah, I thought about it when I kept adding more 😅 I think we could do it separately so this PR doesn't get bigger than it already is.

File: ReactApp/src/components/Editor.jsx
Comment (geriux) [reply ReactApp/src/components/Editor.jsx:66]: It looks like `window.editor` is not used at all, so we could just remove it.

Comment (geriux) [reply ReactApp/src/components/Editor.jsx:66]: Oh ok! I see we are setting some callbacks like `editor.setContent`. These could be also set once and use `window.editor` directly if needed.

Let's do this in a separate PR that can improve how the editor communicates with the host apps.

File: ReactApp/src/misc/api-fetch-setup.js
Comment (geriux) [reply ReactApp/src/misc/api-fetch-setup.js:95]: By just having the API credentials set, as soon as the editor loads with these changes I'm seeing failures to:

`https://public-api.wordpress.com/wp/v2/sites/SITE_ID/sites/SITE_ID/blocks?per_page=100&_locale=user`

It might be an issue related to how the reusable blocks are fetched, but I'm not sure if it would happen with other endpoints. I'd say to include this check. What do you think?

Comment (dcalhoun) [reply ReactApp/src/misc/api-fetch-setup.js:95]: Interesting. I'm curious as to why, but I can research further later. Yes, I agree including the check makes sense.

File: ReactApp/src/components/hooks/use-editor-styles.js
Comment (geriux) [reply ReactApp/src/components/hooks/use-editor-styles.js]: Updated in https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/b6560a8b2b97622ac67033ff51f1d37b3c4e3df3

File: ReactApp/src/main.jsx
Comment (geriux) [reply ReactApp/src/main.jsx]: Updated in https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/c4426ba888c964e4122551a27638ee36a3adbf8f

File: ReactApp/src/misc/Helpers.js
Comment (geriux) [reply ReactApp/src/misc/Helpers.js]: Updated in https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/54aedfb4541fb8465a6b4caa2c2c58f1d86334a3

File: ReactApp/src/misc/Helpers.js
Comment (geriux) [reply ReactApp/src/misc/Helpers.js]: Updated in https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/64665da320dd26e6a8999539340b1d73794b1442 thanks!

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (geriux) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: I've updated it in this commit https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/54aedfb4541fb8465a6b4caa2c2c58f1d86334a3

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (geriux) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: I've changed the approach in https://github.com/wordpress-mobile/GutenbergKit/pull/21/commits/54aedfb4541fb8465a6b4caa2c2c58f1d86334a3 thanks!

**Comment by geriux** [comment]
> Great work! This is a lot of crucial functionality that is incredibly valuable to merge.

Thank you so much for the review and testing! 🚀 

> Testing went relatively well for me. The two things I noted:
> 
> 1. Selecting media from Media Library does not work for me; tapping the button doesn't display anything.

I forgot to mention this. Unfortunately, there's no built-in support for the Media Library within the Web Editor as it relies on the WordPress Media Library functionality.

I think we could have a hybrid approach to open the native Media Library picker. I'll create a GitHub issue for this.

> 2. Media uploaded to private sites results in broken images within the editor.

I'll create a ticket for this as well. I think currently this is solved by using the cookies which we could try to set to the WebView but it needs more investigation as using cookies might change anytime.

**Comment by dcalhoun** [comment]
> I forgot to mention this. Unfortunately, there's no built-in support for the Media Library within the Web Editor as it relies on the WordPress Media Library functionality.

Aha! My confusion is a result of me reading the testing instructions too quickly and _not_ watching your example videos. 😄 I now understand the media tests were solely for the inserter _Media_ tab. 

> I'll create a ticket for this [broken media on private sites] as well. I think currently this is solved by using the cookies which we could try to set to the WebView but it needs more investigation as using cookies might change anytime.

Good to know. My understanding is that cookie authentication is a non-starter for this product moving forward. We'll likely need to further consider approaches for fetching private media. 🤔

**Comment by geriux** [comment]
> Good to know. My understanding is that cookie authentication is a non-starter for this product moving forward. We'll likely need to further consider approaches for fetching private media. 🤔

Yeah, let's work on that next 👍

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: Noting that we could still simplify `hasThemeStylesEnabled`, if desired.

**Comment by geriux** [comment]
I'll merge this now, if there's more feedback I can happily address it in new PRs. Thanks again for the review!

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: After reviewing your changes here, I now understand your original let guard clause. Feel free to revert the changes for these specific lines if you prefer the previous implementation. Either approach is fine with me.

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (geriux) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: Oh I missed that, I'll test that suggestion. I was trying to avoid setting a value wrapped with `Optional()`. I'll check if that handles it.

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (geriux) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: I think with the latest changes its ok because we are checking `id != nil` before setting the post data, and then unwrapping the value with `id!`

File: Sources/GutenbergKit/Sources/EditorViewController.swift
Comment (dcalhoun) [reply Sources/GutenbergKit/Sources/EditorViewController.swift]: Gotcha. It is very possible my suggestion is invalid. Feel free to disregard.

**Pull request merged**
